### PR TITLE
fix(ansible): forward `-- <args>` passthrough to ansible-playbook

### PR DESCRIPTION
**Pull request by mtb-xt**
## what

- Fix `atmos ansible playbook <component> -s <stack> -- <ansible-args>` failing with a misleading `Unknown command \`<component>\`` error.
- `playbookCmd` now uses a separator-aware argument validator instead of `cobra.ExactArgs(1)`: it requires exactly one positional component, counting only the arguments **before** a `--` separator.
- Adds regression tests for the `--` passthrough path.

## why

- `cobra.ExactArgs(1)` counts every positional argument, including the ansible-playbook passthrough tokens captured after `--`. Any invocation using `--` (e.g. `-- --check`, `-- --tags web`, `-- --limit host`) therefore tripped argument validation with 2+ "positional" args, and the root `UsageFunc` rendered that failure as an `Unknown command` error naming the component.
- This made it impossible to forward native `ansible-playbook` flags, even though the command help documents `[options]` and points at the ansible-playbook CLI docs, and the executor (`buildCommandArgs`) already appends `info.AdditionalArgsAndFlags` to the command.
- `cobra.ArgsLenAtDash()` returns `-1` when no `--` is present, so the validator still checks the full slice for the common case and keeps the original "exactly one component" contract.

### Before

```console
$ atmos ansible playbook myapp -s prod -- --check
Error: Unknown command `myapp` for `atmos ansible playbook`
```

### After

```console
$ atmos ansible playbook myapp -s prod --dry-run -- --check
ansible-playbook --extra-vars @/.../prod-myapp.ansible.vars.yaml -i inventory/hosts.yml --check playbook.yml
```

## references

- No existing issue; reproducible on every release that ships `atmos ansible playbook` (verified on v1.220.0).
- Ansible check mode and other passthrough flags: https://docs.ansible.com/ansible/latest/cli/ansible-playbook.html

## Summary by CodeRabbit

* **New Features**
  * Forward additional args after `--` verbatim to the underlying playbook command; positional validation ignores these passthrough args.

* **Bug Fixes**
  * Require exactly one component before `--` and return clear validation errors when missing or duplicated.

* **Tests**
  * Added comprehensive tests for separator-aware parsing, validation, and passthrough behavior, including prompt-aware validation scenarios.

### CODE REVIEW

File: cmd/ansible/ansible_test.go [hunk 1]
```diff
@@ -461,6 +461,49 @@ func TestPlaybookCmdArgsValidation(t *testing.T) {
...
+
+		// Args validator passes for the parsed args.
+		assert.NoError(t, c.Args(c, c.Flags().Args()))
+	})
+
+	t.Run("rejects missing component before --", func(t *testing.T) {
+		c := newCmd()
+		c.SetArgs([]string{"-s", "my-stack", "--", "--check"})
+		assert.Error(t, c.Execute())
+	})
```
[the remark this answers is not in the document]
Comment (mtb-xt) [reply cmd/ansible/ansible_test.go:507]: Good call — addressed in 6b2f5bc5. Both negative cases now assert the specific wrapped sentinel via `assert.ErrorIs(t, err, errUtils.ErrInvalidComponentArgument)` (used `assert.ErrorIs` instead of importing `errors`). Verified the sentinel propagates through `c.Execute()`; `cmd/ansible` tests pass.

File: cmd/ansible/ansible_test.go [hunk 2]
```diff
@@ -461,6 +461,49 @@ func TestPlaybookCmdArgsValidation(t *testing.T) {
...
+
+	t.Run("rejects two components before --", func(t *testing.T) {
+		c := newCmd()
+		c.SetArgs([]string{"comp1", "comp2", "--", "--check"})
+		assert.Error(t, c.Execute())
+	})
```
[the remark this answers is not in the document]
Comment (mtb-xt) [reply cmd/ansible/ansible_test.go]: Good call — addressed in 6b2f5bc5. Both negative cases now assert the specific wrapped sentinel via `assert.ErrorIs(t, err, errUtils.ErrInvalidComponentArgument)` (used `assert.ErrorIs` instead of importing `errors`). Verified the sentinel propagates through `c.Execute()`; `cmd/ansible` tests pass.

**Comment by osterman** [comment]
@coderabbitai are we sure this fix is not curve fitting? The problem I see is that it's in the flag handler itself. And that the args passed in both in the original code (which had exact) and in the patch, do not get their args from the flag handler. I'm pretty sure we have a function for this use case. Can you reference our flag handler agent/skills and compared to other commands using the flag handler. Either way the fix belongs in flag handler if there is not already a fix there.

**Comment by mtb-xt** [comment]
@osterman you're right — thanks for catching it. The first patch only fixed the `Args` validator; the actual arg *ingestion* still bypassed the separator. Fixed in `81ae1288`.

**What was wrong:** `cmd/ansible/ansible.go`'s `initConfigAndStacksInfo` called `processArgs(args)` → `args[1:]` to build `AdditionalArgsAndFlags`. That has no concept of `--`; the passthrough landing correctly was coincidental (cobra happens to pass `["component", "--check"]` flat, so `args[1:]` got it by accident).

**Fix:** ingestion now derives the split from Cobra's `ArgsLenAtDash()` — the same authoritative separator source used by `cmd/auth/shell.go` (`getSeparatedArgs`) and `cmd/auth/exec.go`. Component comes from the positional args *before* `--`; everything *after* `--` is forwarded verbatim to `ansible-playbook`. No-`--` behavior is unchanged.

**On reusing `ParseSeparatedArgs` directly:** `cmd/flag_utils.go`'s `ParseSeparatedArgs`/`SeparatedArgs` live in `package cmd`, and `cmd/root.go` imports `cmd/ansible` — so `cmd/ansible` importing `package cmd` is an import cycle (same reason the tests note `cmd.NewTestKit` can't be used here). So I followed the identical `ArgsLenAtDash()` approach those `auth` subcommands use locally, rather than the literal helper. **If you'd prefer the canonical util reused everywhere, I'm happy to promote `ParseSeparatedArgs`/`SeparatedArgs` into a shared package (e.g. `pkg/flags`) and have `cmd`, `cmd/auth`, and `cmd/ansible` all consume it** — that'd remove the duplication in `auth` too. Your call on scope.

The separator-aware `Args` validator (`validatePlaybookArgs`) stays — cobra's `ExactArgs(1)` counts post-`--` tokens and rejects valid invocations.

Added `TestInitConfigAndStacksInfoDoubleDash`, which parses a real argv (so `ArgsLenAtDash` is populated) and asserts the passthrough is sourced from the separator, not a positional offset. `cmd/ansible` + `pkg/component/ansible` tests pass.

**Comment by mtb-xt** [comment]
@osterman I started moving this to the flag handler (using #2595's `devcontainer exec` as the template) and hit a constraint worth your call before I refactor further. Sharing the evidence.

### The handler only captures flag-like passthrough with `DisableFlagParsing=true`

Cobra consumes the `--` before `RunE` runs, so the args a command receives no longer contain the literal `--` (only `ArgsLenAtDash()` records where it was):

```
input:  playbook comp -s stk -- --check --tags net
RunE args:  ["comp" "--check" "--tags" "net"]   ArgsLenAtDash=1
```

If I then hand those args to `StandardFlagParser.Parse()`, it re-parses them in a fresh combined FlagSet and there is no `--` left, so the passthrough flags are treated as unknown command flags. Probe with the real parser (component positional + a `-p` flag):

```
DisableFlagParsing=false:  Parse("comp -p site.yml -- --check --tags net")
  -> parseErr: failed to parse flags: unknown flag: --check     (posn=[] sep=[])

DisableFlagParsing=true:   RunE gets raw args incl. "--"
  -> posn=["comp"]  sep=["--check" "--tags" "net"]              ✅
```

So for ansible `playbook <component> -- <ansible flags>` to flow through `Parse()`, the command needs `DisableFlagParsing=true` **and** every flag read (`--stack`, `--dry-run`, `--identity`, `--playbook`, `--inventory`, globals) has to move off `cmd.Flag(...)` onto the parsed result. Happy to do that — it's just a bigger, more invasive change than the current patch, and it touches the ansible-specific bits (the reserved `-i` = inventory vs `--identity` long-form handling), so I wanted to confirm the direction first.

### Heads-up on #2595

By the same mechanism, `devcontainer exec` in #2595 doesn't set `DisableFlagParsing` (checked every file in the package at the PR head, including the parent command), and its `parseExecInvocation` keys off `slices.Contains(args, "--")` — but `args` in `RunE` never contains `--` (cobra strips the first one). So `exec <name> -- <cmd>` takes the no-separator branch, and `exec <name> -- docker run --rm alpine` would fail in `Parse()` with `unknown flag: --rm` (likewise `-- ls -la` → `unknown shorthand flag: 'l' in -la`).

**Why CI doesn't catch it:** the tests in `parser_test.go` call `parseExecInvocation(args, parsed)` with hand-built slices that still contain the literal `"--"` (e.g. `{"app", "--", "docker", "run", "--rm", "--pull=never", "alpine"}`) and feed the parser directly via `parseTestArgs` — they never route through `cobra.Command.Execute()`. That args shape can't occur in the real CLI path, so the green tests are validating an unreachable input. An `Execute()`-level test (`cmd.SetArgs(...)` + `cmd.Execute()`) would reproduce the failure.

(One nit for completeness: a *second* `--` survives cobra's stripping, so `slices.Contains` can fire in that edge case — with garbled semantics either way.)

### On the CodeRabbit/adversarial concern about `ArgsLenAtDash` statefulness

For the one-shot CLI it's stateless (one parse per process), and the handler itself reads `ArgsLenAtDash()` too — just on a fresh FlagSet per `Parse()`. So it's not a production risk either way; the handler path is cleaner regardless.

### Which way do you want it?

1. **Full handler migration** — `DisableFlagParsing=true` on `playbook`, component positional via `NewPositionalArgsBuilder`, all flags via `Parse()`. (I'll also propose the `DisableFlagParsing` fix for #2595's exec if you want.) Note `DisableFlagParsing` also costs cobra's auto-`--help` handling and changes completion behavior — solvable, just part of the scope.
2. **Keep the current `ArgsLenAtDash()` split** in the command (verified forwarding `-- --check`, `-- --tags ...` end-to-end) as a smaller fix.

I'll go whichever way you prefer.

**Comment by osterman** [comment]
> The handler only captures flag-like passthrough with DisableFlagParsing=true

@coderabbitai this is not my recollection. Your suggestion corresponds to how things worked before we introduced the flag-handler. There's a big reason why we wanted to introduce the Flag Handler, because by disabling flag parsing, meant we kept missing problems with our flag parsing. Instead, we wanted Cobra to validate all the flags, so we implemented a preprocessor with a Flag Handler that should have normalized/sanitized all the flags so that they work and are fully validated by Cobra. We then integrated that together with the compatibility flags, as well as handling the end of args (`--`) marker. If disabling flag parsing is required, it should be entirely handled by the flag handler and not something that the developer makes a decision on.

**Comment by osterman** [comment]
Thanks for the fix and the regression tests — the bug is real and your root-cause analysis (Cobra's `ExactArgs(1)` counting post-`--` args) is correct. But the implementation needs to move: as written it adds a 5th copy of `--` splitting logic to the codebase, hand-rolled inside `cmd/ansible/`. Separator handling is determined by the flag handler (`pkg/flags/`), not per-command code.

Two facts from the existing infrastructure:

1. `pkg/flags/standard.go` already does this split. `StandardFlagParser.extractArgs()` separates `ParsedConfig.PositionalArgs` (pre-`--`) from `SeparatedArgs` (post-`--`), and `validatePositionalArgs()` validates only the pre-dash args. The ansible command just doesn't route through it — `processArgs()` splits on positional offset (`args[1:]`), which is the actual bug you found.
2. The real infrastructure gap is in `pkg/flags/positional_args_builder.go`: the validators `generateValidator()` emits (`cobra.ExactArgs` etc.) are not separator-aware when used as `cmd.Args`. That's the bug *class* — any passthrough command using a builder-generated validator (or raw `ExactArgs`) hits it. Fix it there once and every command gets `--` handling for free.

Here's the shape of the fix we'd like:

**`pkg/flags/separator.go` (new):**

```go
// SplitArgsAtDash splits args into positional args (before "--") and
// pass-through args (after "--") using Cobra's ArgsLenAtDash().
func SplitArgsAtDash(cmd *cobra.Command, args []string) (positional, separated []string)

// SeparatorAwareValidator wraps a cobra.PositionalArgs validator so it
// validates only the args before the "--" separator.
func SeparatorAwareValidator(validator cobra.PositionalArgs) cobra.PositionalArgs
```

**`pkg/flags/positional_args_builder.go`:** wrap the validators returned by `generateValidator()` (and the count inside `GeneratePromptAwareValidator()`) with the separator-aware logic.

**`cmd/ansible/playbook.go`:** drop `cobra.ExactArgs(1)` and `validatePlaybookArgs`; declare the contract through the builder:

```go
argsBuilder := flags.NewPositionalArgsBuilder()
argsBuilder.AddArg(&flags.PositionalArgSpec{
    Name:        "component",
    Description: "Ansible component",
    Required:    true,
    TargetField: "Component",
})
_, validator, usage := argsBuilder.Build()
playbookCmd.Use = "playbook " + usage
playbookCmd.Args = validator
```

**`cmd/ansible/ansible.go`:** replace `processArgs`/`splitComponentAndPassthrough` with the shared helper:

```go
positional, separated := flags.SplitArgsAtDash(cmd, args)
if len(positional) > 0 {
    info.ComponentFromArg = positional[0]
}
if len(positional) > 1 {
    info.AdditionalArgsAndFlags = positional[1:] // Preserve no-dash fallback.
}
info.AdditionalArgsAndFlags = append(info.AdditionalArgsAndFlags, separated...)
```

**Error semantics:** drop the `ErrInvalidComponentArgument` wrapping — the builder-generated validator produces the standard `accepts 1 arg(s), received N` message consistent with every other command.

**Tests:** keep your regression tests (adapted), and add unit tests in `pkg/flags` for `SplitArgsAtDash` + `SeparatorAwareValidator` (no dash, dash at position 0, dash with real parsed `cobra.Command`, nil cmd).

Out of scope for this PR (we'll consolidate separately): migrating the existing duplicates (`cmd/flag_utils.go` `ParseSeparatedArgs`, `cmd/args_separator.go` `ExtractSeparatedArgs`, `cmd/auth/{shell,exec}.go` helpers) onto the new helper, and the eventual move of ansible onto `parser.Parse(ctx, args)` like `cmd/terraform/backend/backend_describe.go`.

**Comment by mtb-xt** [comment]
@osterman Implemented exactly per your spec in 09a3d51d — thanks for the precise direction, and you're right that my `DisableFlagParsing` claim described the pre-flag-handler workaround, not the architecture. The earlier hand-rolled splitting in `cmd/ansible` is gone.

**`pkg/flags/separator.go` (new):** `SplitArgsAtDash(cmd, args)` and `SeparatorAwareValidator(validator)` as specified. Nil cmd, no separator, and an out-of-range dash index (caller passes a shorter slice than Cobra parsed) all degrade to "everything is positional".

**`pkg/flags/positional_args_builder.go`:** `generateValidator()` wraps its `ExactArgs`/`MaximumNArgs`/`RangeArgs` returns with `SeparatorAwareValidator`, and the count inside `GeneratePromptAwareValidator()` now uses the pre-dash positional count (its `standardValidator` delegate is already wrapped). Existing builder consumers (`cmd/theme`, `cmd/terraform/backend/*`) pass their suites unchanged — their tests invoke validators with nil cmd, where the wrapper is a no-op by design.

**`cmd/ansible/playbook.go`:** contract declared through `NewPositionalArgsBuilder` (required `component`); `Use` is now the generated `playbook <component>`. `validatePlaybookArgs` and the `ErrInvalidComponentArgument` wrapping are deleted — errors are the standard `accepts 1 arg(s), received N`.

**`cmd/ansible/ansible.go`:** `processArgs`/`splitComponentAndPassthrough` replaced with your exact `flags.SplitArgsAtDash` ingestion shape (no-dash fallback preserved, separated args appended).

**Tests:** `pkg/flags/separator_test.go` covers your requested matrix (no dash, dash at position 0, dash with a real parsed `cobra.Command`, nil cmd, plus the short-slice guard) and the builder validators (required/optional/prompt-aware). The ansible regression tests are adapted to the builder validator and standard error text.

Verified end-to-end on the built binary: `playbook <component> -s <stack> -- --check --tags net --limit web` forwards to `ansible-playbook`, no-separator invocations are unchanged, and missing/surplus component errors render consistently with other commands.

Left out per your scope note: consolidating `cmd/flag_utils.go` / `cmd/args_separator.go` / `cmd/auth/{shell,exec}.go` onto the new helper, and the `parser.Parse(ctx, args)` migration. Happy to pick either up as a follow-up if useful.

**Comment by mtb-xt** [comment]
Heads-up on the red `autofix` check: it's a fork-PR infrastructure limitation, not a formatting issue in this PR.

- The failing step is `atmos pro commit -m "[autocommit] formatting fixes" --all`, which dies with `not running in GitHub Actions or missing OIDC token environment`. Fork-PR workflow runs never get an OIDC token (GitHub doesn't grant `id-token` to forks), so `atmos pro commit` can't authenticate on any external contribution.
- There was nothing for it to commit anyway: I ran the same fixer the job runs (`gofumpt -l -w` over the exact changed-file list from the job log: `pkg/flags/separator.go`, `pkg/flags/separator_test.go`, `cmd/ansible/{ansible,ansible_test,playbook}.go`, `pkg/flags/positional_args_builder.go`) — zero diffs, whitespace/EOF clean.

So every fork PR will show this check red once workflows are approved. Possibly worth guarding the commit step on `github.event.pull_request.head.repo.fork == false` (or skipping commit when the tree is clean) on the Atmos Pro side.

Acceptance tests are still queued — happy to address anything they surface.

**Review by osterman — APPROVED** [review]

**Review by aknysh — APPROVED** [review]
thanks @mtb-xt
